lib/spack/spack/util/windows_registry.py

```diff
@@ -218,30 +244,47 @@ def get_subkeys(self):
         with self.invalid_reg_ref_error_handler():
             return self.reg.subkeys
 
+    def get_matching_subkeys(self, subkey_name):
+        self._regex_match_subkeys(subkey_name)
+
     def get_values(self):
         if not self._valid_reg_check():
             raise RegistryError("Cannot query values from invalid key %s" % self.key)
         with self.invalid_reg_ref_error_handler():
             return self.reg.values
 
-    def _traverse_subkeys(self, stop_condition):
+    def _traverse_subkeys(self, stop_condition, collect_found=False):
         """Perform simple BFS of subkeys, returning the key
         that successfully triggers the stop condition.
         Args:
             stop_condition: lambda or function pointer that takes a single argument
                             a key and returns a boolean value based on that key
+            collect_found: boolean value, if True, the traversal collects and returns
+                            all keys meeting stop condition. If false, once stop
+                            condition is met, the key that triggered the condition '
+                            is returned.
         Return:
             the key if stop_condition is triggered, or None if not
         """
+        collection = []
         if not self._valid_reg_check():
             raise RegistryError("Cannot query values from invalid key %s" % self.key)
         with self.invalid_reg_ref_error_handler():
             queue = self.reg.subkeys
             for key in queue:
                 if stop_condition(key):
-                    return key
+                    if collect_found:
+                        collection.append(key)
+                    else:
+                        return key
                 queue.extend(key.subkeys)
-            return None
+            return collection if collection else None
+
+    def _find_subkey_s(self, subkey_name, condition, recursive=True, collect_found=False):
```

Comment:
What does the `_s` at the end mean? If it means nothing, then IMO `_helper` would be more useful.

Also, it seems that `recursive` and `condition` are redundant: `condition` is always set when `recursive=True` (also might be an opportunity to choose a name with more semantic value than "recursive"), e.g.

```
# I also suggest enforcing that exactly one of the two args is set
_find_subkey_s(self, direct_subkey=None, search_key=None, collect_found=False):
  if direct_subkey and ssearch_key:
    raise Exception(cannot set both)
  elif not (direct_subkey or search_key):
    raise...
```

Reply:
I intended it to indicate an optional plural like "option(s)" to indicate the method may return one or more reg keys. 

I agree with the suggestion overall, will apply with my next push.